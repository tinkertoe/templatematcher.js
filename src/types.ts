export interface Point {
  x: number
  y: number
}

export interface Match {
  location: Point
  averageSimilarity: number,
  allSimilarities?: number[],
  timestamp?: number
}

export interface ImageData {
  data: Uint8ClampedArray,
  width: number,
  height: number
}

