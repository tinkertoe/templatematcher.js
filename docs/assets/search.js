window.searchData = JSON.parse("{\"kinds\":{\"128\":\"Class\",\"256\":\"Interface\",\"512\":\"Constructor\",\"1024\":\"Property\",\"2048\":\"Method\"},\"rows\":[{\"kind\":128,\"name\":\"TemplateMatcher\",\"url\":\"classes/TemplateMatcher.html\",\"classes\":\"tsd-kind-class\"},{\"kind\":2048,\"name\":\"downloadTexture\",\"url\":\"classes/TemplateMatcher.html#downloadTexture\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":512,\"name\":\"constructor\",\"url\":\"classes/TemplateMatcher.html#constructor\",\"classes\":\"tsd-kind-constructor tsd-parent-kind-class\",\"parent\":\"TemplateMatcher\"},{\"kind\":1024,\"name\":\"gpu\",\"url\":\"classes/TemplateMatcher.html#gpu\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":1024,\"name\":\"computeSimilarities\",\"url\":\"classes/TemplateMatcher.html#computeSimilarities\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":1024,\"name\":\"computeAverageSimilarities\",\"url\":\"classes/TemplateMatcher.html#computeAverageSimilarities\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":1024,\"name\":\"computeHighestAverages\",\"url\":\"classes/TemplateMatcher.html#computeHighestAverages\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":1024,\"name\":\"computeSingleLocation\",\"url\":\"classes/TemplateMatcher.html#computeSingleLocation\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":2048,\"name\":\"highestSimilarities\",\"url\":\"classes/TemplateMatcher.html#highestSimilarities\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"TemplateMatcher\"},{\"kind\":2048,\"name\":\"highestSimilarity\",\"url\":\"classes/TemplateMatcher.html#highestSimilarity\",\"classes\":\"tsd-kind-method tsd-parent-kind-class\",\"parent\":\"TemplateMatcher\"},{\"kind\":2048,\"name\":\"filteredSimilarities\",\"url\":\"classes/TemplateMatcher.html#filteredSimilarities\",\"classes\":\"tsd-kind-method tsd-parent-kind-class\",\"parent\":\"TemplateMatcher\"},{\"kind\":2048,\"name\":\"verifyMatch\",\"url\":\"classes/TemplateMatcher.html#verifyMatch\",\"classes\":\"tsd-kind-method tsd-parent-kind-class\",\"parent\":\"TemplateMatcher\"},{\"kind\":2048,\"name\":\"destroy\",\"url\":\"classes/TemplateMatcher.html#destroy\",\"classes\":\"tsd-kind-method tsd-parent-kind-class\",\"parent\":\"TemplateMatcher\"},{\"kind\":256,\"name\":\"Point\",\"url\":\"interfaces/Point.html\",\"classes\":\"tsd-kind-interface\"},{\"kind\":1024,\"name\":\"x\",\"url\":\"interfaces/Point.html#x\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Point\"},{\"kind\":1024,\"name\":\"y\",\"url\":\"interfaces/Point.html#y\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Point\"},{\"kind\":256,\"name\":\"Match\",\"url\":\"interfaces/Match.html\",\"classes\":\"tsd-kind-interface\"},{\"kind\":1024,\"name\":\"location\",\"url\":\"interfaces/Match.html#location\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Match\"},{\"kind\":1024,\"name\":\"averageSimilarity\",\"url\":\"interfaces/Match.html#averageSimilarity\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Match\"},{\"kind\":1024,\"name\":\"allSimilarities\",\"url\":\"interfaces/Match.html#allSimilarities\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Match\"},{\"kind\":1024,\"name\":\"timestamp\",\"url\":\"interfaces/Match.html#timestamp\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Match\"},{\"kind\":256,\"name\":\"ImageData\",\"url\":\"interfaces/ImageData.html\",\"classes\":\"tsd-kind-interface\"},{\"kind\":1024,\"name\":\"data\",\"url\":\"interfaces/ImageData.html#data\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"ImageData\"},{\"kind\":1024,\"name\":\"width\",\"url\":\"interfaces/ImageData.html#width\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"ImageData\"},{\"kind\":1024,\"name\":\"height\",\"url\":\"interfaces/ImageData.html#height\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"ImageData\"}],\"index\":{\"version\":\"2.3.9\",\"fields\":[\"name\",\"comment\"],\"fieldVectors\":[[\"name/0\",[0,28.526]],[\"comment/0\",[]],[\"name/1\",[1,28.526]],[\"comment/1\",[]],[\"name/2\",[2,28.526]],[\"comment/2\",[]],[\"name/3\",[3,28.526]],[\"comment/3\",[]],[\"name/4\",[4,28.526]],[\"comment/4\",[]],[\"name/5\",[5,28.526]],[\"comment/5\",[]],[\"name/6\",[6,28.526]],[\"comment/6\",[]],[\"name/7\",[7,28.526]],[\"comment/7\",[]],[\"name/8\",[8,28.526]],[\"comment/8\",[]],[\"name/9\",[9,28.526]],[\"comment/9\",[]],[\"name/10\",[10,28.526]],[\"comment/10\",[]],[\"name/11\",[11,28.526]],[\"comment/11\",[]],[\"name/12\",[12,28.526]],[\"comment/12\",[]],[\"name/13\",[13,28.526]],[\"comment/13\",[]],[\"name/14\",[14,28.526]],[\"comment/14\",[]],[\"name/15\",[15,28.526]],[\"comment/15\",[]],[\"name/16\",[16,28.526]],[\"comment/16\",[]],[\"name/17\",[17,28.526]],[\"comment/17\",[]],[\"name/18\",[18,28.526]],[\"comment/18\",[]],[\"name/19\",[19,28.526]],[\"comment/19\",[]],[\"name/20\",[20,28.526]],[\"comment/20\",[]],[\"name/21\",[21,28.526]],[\"comment/21\",[]],[\"name/22\",[22,28.526]],[\"comment/22\",[]],[\"name/23\",[23,28.526]],[\"comment/23\",[]],[\"name/24\",[24,28.526]],[\"comment/24\",[]]],\"invertedIndex\":[[\"allsimilarities\",{\"_index\":19,\"name\":{\"19\":{}},\"comment\":{}}],[\"averagesimilarity\",{\"_index\":18,\"name\":{\"18\":{}},\"comment\":{}}],[\"computeaveragesimilarities\",{\"_index\":5,\"name\":{\"5\":{}},\"comment\":{}}],[\"computehighestaverages\",{\"_index\":6,\"name\":{\"6\":{}},\"comment\":{}}],[\"computesimilarities\",{\"_index\":4,\"name\":{\"4\":{}},\"comment\":{}}],[\"computesinglelocation\",{\"_index\":7,\"name\":{\"7\":{}},\"comment\":{}}],[\"constructor\",{\"_index\":2,\"name\":{\"2\":{}},\"comment\":{}}],[\"data\",{\"_index\":22,\"name\":{\"22\":{}},\"comment\":{}}],[\"destroy\",{\"_index\":12,\"name\":{\"12\":{}},\"comment\":{}}],[\"downloadtexture\",{\"_index\":1,\"name\":{\"1\":{}},\"comment\":{}}],[\"filteredsimilarities\",{\"_index\":10,\"name\":{\"10\":{}},\"comment\":{}}],[\"gpu\",{\"_index\":3,\"name\":{\"3\":{}},\"comment\":{}}],[\"height\",{\"_index\":24,\"name\":{\"24\":{}},\"comment\":{}}],[\"highestsimilarities\",{\"_index\":8,\"name\":{\"8\":{}},\"comment\":{}}],[\"highestsimilarity\",{\"_index\":9,\"name\":{\"9\":{}},\"comment\":{}}],[\"imagedata\",{\"_index\":21,\"name\":{\"21\":{}},\"comment\":{}}],[\"location\",{\"_index\":17,\"name\":{\"17\":{}},\"comment\":{}}],[\"match\",{\"_index\":16,\"name\":{\"16\":{}},\"comment\":{}}],[\"point\",{\"_index\":13,\"name\":{\"13\":{}},\"comment\":{}}],[\"templatematcher\",{\"_index\":0,\"name\":{\"0\":{}},\"comment\":{}}],[\"timestamp\",{\"_index\":20,\"name\":{\"20\":{}},\"comment\":{}}],[\"verifymatch\",{\"_index\":11,\"name\":{\"11\":{}},\"comment\":{}}],[\"width\",{\"_index\":23,\"name\":{\"23\":{}},\"comment\":{}}],[\"x\",{\"_index\":14,\"name\":{\"14\":{}},\"comment\":{}}],[\"y\",{\"_index\":15,\"name\":{\"15\":{}},\"comment\":{}}]],\"pipeline\":[]}}");