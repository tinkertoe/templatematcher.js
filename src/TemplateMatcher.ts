import { GPU, IKernelRunShortcut, Texture } from 'gpu.js'
import { Match, ImageData } from './types'

export class TemplateMatcher {
  private gpu: GPU
  private computeSimilarities: IKernelRunShortcut
  private computeAverageSimilarities: IKernelRunShortcut
  private computeHighestAverages: IKernelRunShortcut
  private computeSingleLocation: IKernelRunShortcut

  constructor () {
    this.gpu = new GPU()

    this.computeSimilarities = this.gpu.createKernel(function (
      imageData: number[],
      imageWidth: number,
      templateData: number[],
      templateWidth: number
    ): number {
      // Compute similarity value for every template pixel at all overlays

      // Template pixel values
      // Take steps in whole pixels (4x) and add channel offset
      const templateRed = templateData[4 * this.thread.x + 0]
      const templateGreen = templateData[4 * this.thread.x + 1]
      const templateBlue = templateData[4 * this.thread.x + 2]
    
      // Calculate coordinates for underlying pixel on image
      // Take template overlay origin and add x/y component of current template pixel
      const imageX = this.thread.y + (this.thread.x % templateWidth)
      const imageY = this.thread.z + Math.trunc(this.thread.x / templateWidth)
    
      // Coresponding image pixel values
      // Take steps in whole pixels (4x) and add channel offset
      const imageRed = imageData[4 * (imageY*imageWidth + imageX) + 0]
      const imageGreen = imageData[4 * (imageY*imageWidth + imageX) + 1]
      const imageBlue = imageData[4 * (imageY*imageWidth + imageX) + 2]
      
      // Difference => Similarity => Percentage
      const pixelDifference = (Math.abs(templateRed - imageRed) + Math.abs(templateGreen - imageGreen) + Math.abs(templateBlue - imageBlue)) / 3
      const pixelSimilarity = 255 - pixelDifference
      return pixelSimilarity / 255 * 100 // Percentage
    }, {
      dynamicOutput: true,
      pipeline: true
    })
    
    this.computeAverageSimilarities = this.gpu.createKernel(function (
      similarities: number[][][],
      similarityCount: number
    ) {
      // Compute average similarity value for each pixel on image
      let similaritySum = 0
      for (let z = 0; z < similarityCount; z++) {
        similaritySum += similarities[this.thread.y][this.thread.x][z]
      }
      return similaritySum / similarityCount
    }, {
      dynamicOutput: true,
      pipeline: true
    })

    this.computeHighestAverages = this.gpu.createKernel(function (
      averageSimilarities: number[][],
      rowLength: number
    ) {
      const y = this.thread.x
    
      let highestSimilarity = -1
      let highestSimilarityX = 0
    
      for (let x = 0; x < rowLength; x++) {
        if (averageSimilarities[y][x] > highestSimilarity) {
          highestSimilarity = averageSimilarities[y][x]
          highestSimilarityX = x
        }
      }
    
      return [highestSimilarity, highestSimilarityX]
    }, {
      dynamicOutput: true,
      pipeline: false
    })

    this.computeSingleLocation = this.gpu.createKernel(function (
      imageData: number[],
      imageWidth: number,
      templateData: number[],
      templateWidth: number,
      locationX: number,
      locationY: number
    ) {
      // Template pixel values
      // Take steps in whole pixels (4x) and add channel offset
      const templateRed = templateData[4 * this.thread.x + 0]
      const templateGreen = templateData[4 * this.thread.x + 1]
      const templateBlue = templateData[4 * this.thread.x + 2]
    
      // Calculate coordinates for underlying pixel on image
      // Take template overlay origin and add x/y component of current template pixel
      const imageX = locationX + (this.thread.x % templateWidth)
      const imageY = locationY + Math.trunc(this.thread.x / templateWidth)
    
      // Coresponding image pixel values
      // Take steps in whole pixels (4x) and add channel offset
      const imageRed = imageData[4 * (imageY*imageWidth + imageX) + 0]
      const imageGreen = imageData[4 * (imageY*imageWidth + imageX) + 1]
      const imageBlue = imageData[4 * (imageY*imageWidth + imageX) + 2]
       
      // Difference => Similarity => Percentage
      const pixelDifference = (Math.abs(templateRed - imageRed) + Math.abs(templateGreen - imageGreen) + Math.abs(templateBlue - imageBlue)) / 3
      const pixelSimilarity = 255 - pixelDifference
      return Math.round(pixelSimilarity / 255 * 100) // Percentage
    }, {
      dynamicOutput: true,
      pipeline: false
    })
  }

  private static async downloadTexture(texture: Texture) {
    return texture.toArray()
  }

  private async highestSimilarities(image: ImageData, template: ImageData) {
    this.computeSimilarities.setOutput([
      template.data.length/4, // this.thread.x: Pixels
      image.width-template.width+1, // this.thread.z: Image width (where template is overlayable)
      image.height-template.height+1 // this.thread.y: Image height (where template is overlayable)
    ])

    const similarities = this.computeSimilarities(image.data, image.width, template.data, template.width) as Texture
    const downloadedSimilarities = TemplateMatcher.downloadTexture(similarities) as Promise<number[][][]>

    this.computeAverageSimilarities.setOutput([
      image.width-template.width+1, // this.thread.y: Image width (where template is overlayable)
      image.height-template.height+1 // this.thread.x: Image height (where template is overlayable)
    ])
    const averageSimilarities = this.computeAverageSimilarities(similarities, template.data.length/4) as Texture

    this.computeHighestAverages.setOutput([
      image.height-template.height+1
    ])
    const highestAverageSimilarities = this.computeHighestAverages(averageSimilarities, image.width-template.width+1) as number[][]

    await downloadedSimilarities // make sure texture is downloaded before it is destroyed

    // Cleanup
    similarities.delete()
    averageSimilarities.delete()

    return {
      downloadedSimilarities,
      highestAverageSimilarities
    }
  }

  public async highestSimilarity(image: ImageData, template: ImageData) {
    const { downloadedSimilarities, highestAverageSimilarities } = await this.highestSimilarities(image, template)

    let highestSimilarity = -Infinity
    let highestSimilarityX = 0
    let highestSimilarityY = 0

    for (let y = 0; y < highestAverageSimilarities.length; y++) {
      if (highestAverageSimilarities[y][0] > highestSimilarity) {
        highestSimilarity = highestAverageSimilarities[y][0]
        highestSimilarityX = highestAverageSimilarities[y][1]
        highestSimilarityY = y
      }
    }

    const match: Match = {
      averageSimilarity: highestSimilarity,
      allSimilarities: (await downloadedSimilarities)[highestSimilarityY][highestSimilarityX],
      location: {
        x: highestSimilarityX,
        y: highestSimilarityY
      }
    }

    return match
  }

  public async filteredSimilarities(image: ImageData, template: ImageData, minSimilarity: number) {
    const { downloadedSimilarities, highestAverageSimilarities } = await this.highestSimilarities(image, template)

    const matches: Match[] = []

    for (let y = 0; y < highestAverageSimilarities.length; y++) {
      if (highestAverageSimilarities[y][0] > minSimilarity) {
        const x = highestAverageSimilarities[y][1]
        matches.push({
          averageSimilarity: highestAverageSimilarities[y][0],
          allSimilarities: (await downloadedSimilarities)[y][x],
          location: { x, y }
        })
      }
    }

    return matches
  }

  public async verifyMatch(image: ImageData, template: ImageData, match: Match) {
    this.computeSingleLocation.setOutput([
      template.data.length/4
    ])
    const similarities = this.computeSingleLocation(image.data, image.width, template.data, template.width, match.location.x, match.location.y) as number[]

    return ( JSON.stringify(similarities) === JSON.stringify(match.allSimilarities) )
  }

  public async destroy() {
    this.computeSimilarities.destroy(true)
    this.computeAverageSimilarities.destroy(true)
    this.computeHighestAverages.destroy(true)
    this.computeSingleLocation.destroy(true)
    this.gpu.destroy()
  }
}