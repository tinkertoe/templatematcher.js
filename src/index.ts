export * from './TemplateMatcher'
export * from './types'