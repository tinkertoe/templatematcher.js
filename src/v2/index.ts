import * as gpu from './gpu/index'
import { imageDataToTexture, chunk } from './gpu/utils'
import { Chunk, Dimensions, Match, Point, WaldoImageData, WaldoTexture } from './types'

type _Match = Match
type _Point = Point
type _WaldoImageData = WaldoImageData
export { _Match as Match, _Point as Point, _WaldoImageData as WaldoImageData }

export class Waldo {
  private gl: WebGLRenderingContext
  private computeSimilarities: gpu.ComputeSimilarities
  private averageSimilarities: gpu.AverageSimilarities
  private findHighestSimilarities: gpu.FindHighestSimilarities
  private findHighestSimilarity: gpu.FindHighestSimilarity
  private downloadTexture: gpu.DownloadTexture
  private _glLocked: boolean = false

  constructor(gl: WebGLRenderingContext) {
    this.gl = gl
    this.gl.getExtension('OES_texture_float')
    
    this.computeSimilarities = new gpu.ComputeSimilarities(this.gl)
    this.averageSimilarities = new gpu.AverageSimilarities(this.gl)
    this.findHighestSimilarities = new gpu.FindHighestSimilarities(this.gl)
    this.findHighestSimilarity = new gpu.FindHighestSimilarity(this.gl)

    this.downloadTexture = new gpu.DownloadTexture(this.gl)
  }

  private lockGl(): Promise<void> {
    // Wait for unlock, then lock
    const poll = (resolve: () => void) => {
      if (!this._glLocked) {
        this._glLocked = true
        resolve()
      } else {
        // GL is locked, retrying in 10ms
        setTimeout(() => { poll(resolve) }, 10)
      }
    }
    return new Promise(poll)
  }

  private unlockGl(): void {
    this._glLocked = false
  }

  private highestSimilarities(imageData: WaldoImageData, templateData: WaldoImageData): Chunk[] {
    // Create textures
    const image = imageDataToTexture(this.gl, imageData)
    const template = imageDataToTexture(this.gl, templateData)

    // Split image into processing chunks that don't exceed the texture size limitation
    const chunks = chunk(image.dimensions, template.dimensions, this.gl.MAX_TEXTURE_SIZE)

    // Do processing for every chunk
    chunks.forEach(chunk => {
      chunk.computedSimilarities = this.computeSimilarities.run(image, template, chunk.region)
      chunk.averagedSimilarities = this.averageSimilarities.run(chunk.computedSimilarities, template.dimensions)
      chunk.highestSimilarities = this.findHighestSimilarities.run(chunk.averagedSimilarities)
    })

    return chunks
  }

  public async highestSimilarity(imageData: WaldoImageData, templateData: WaldoImageData): Promise<Match> {
    await this.lockGl()

    const chunks: Chunk[] = this.highestSimilarities(imageData, templateData)

    let highestSimilarityValue = 0
    let highestSimilarityLocation: Point = { x: 0, y: 0 }

    chunks.forEach(chunk => {
      chunk.highestSimilarity = this.findHighestSimilarity.run(chunk.highestSimilarities as WaldoTexture)
      const { data } = this.downloadTexture.run(chunk.highestSimilarity)

      if (data[0] >= highestSimilarityValue) {
        highestSimilarityValue = data[0]
        highestSimilarityLocation = {
          x: Math.floor(data[1] + chunk.region.origin.x),
          y: Math.floor(data[2] + chunk.region.origin.y)
        }
      }
    })

    this.unlockGl()

    return {
      location: highestSimilarityLocation,
      similarity: highestSimilarityValue
    }
  }

  public async filteredSimilarities(imageData: WaldoImageData, templateData: WaldoImageData, minSimilarity: number): Promise<Match[]> {
    await this.lockGl()

    const chunks: Chunk[] = this.highestSimilarities(imageData, templateData)
    const matches: Match[] = []

    chunks.forEach(chunk => {
      const { data } = this.downloadTexture.run(chunk.highestSimilarities as WaldoTexture)

      for (let i = 0; i < data.length; i = i+4) {
        if (data[i+0] >= minSimilarity) {
          matches.push({
            similarity: data[i+0],
            location: {
              x: Math.floor(data[i+1] + chunk.region.origin.x),
              y: Math.floor(i/4       + chunk.region.origin.y)
            }
          })
        }
      }
    })

    this.unlockGl()
    return matches
  }

  public async verifyMatch(imageData: WaldoImageData, templateData: WaldoImageData, match: Match, similarityMargin: number = 0.05): Promise<boolean> {
    await this.lockGl()

    const image = imageDataToTexture(this.gl, imageData)
    const template = imageDataToTexture(this.gl, templateData)

    const chunk: Chunk = {
      region: {
        origin: {
          x: match.location.x,
          y: match.location.y
        },
        dimensions: { w: 1, h: 1 }
      }
    }

    chunk.computedSimilarities = this.computeSimilarities.run(image, template, chunk.region)
    chunk.averagedSimilarities = this.averageSimilarities.run(chunk.computedSimilarities, template.dimensions)
    const { data } = this.downloadTexture.run(chunk.averagedSimilarities)

    this.unlockGl()
  
    /*
      Check if provided match location is possible!

      Texture lookup at coordinates that are outside of the actual
      texture size return the value of the closest coordinate.
      (Texture doesn't wrap arround but is clamped to the edge)

      Without this check this function could also verify a match that is located outside of the image!
      (As long as the to be verified similarity value is the same as the acutal similarity value on the edge of the image)
    */
    const validDimensions: Dimensions = {
      w: image.dimensions.w - template.dimensions.w + 1,
      h: image.dimensions.h - template.dimensions.h + 1
    }

    if (match.location.x > validDimensions.w || match.location.y > validDimensions.h) {
      return false
    }

    // Verify match if difference between recalculated similarity and specified similarity is within 5% (default)
    return (Math.abs(data[0] - match.similarity) < similarityMargin)
  }

  public async destroy() {
    this.computeSimilarities.destroy()
    this.averageSimilarities.destroy()
    this.findHighestSimilarities.destroy()
    this.findHighestSimilarity.destroy()
    this.downloadTexture.destroy()
    this.gl.getExtension('WEBGL_lose_context')?.loseContext();
    (this.gl.canvas as HTMLCanvasElement)?.remove()
  }
}