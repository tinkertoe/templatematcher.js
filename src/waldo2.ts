import WebGL from 'gl'

import { ImageData } from './types'

export class Waldo {
  private gl: WebGLRenderingContext
  private vertexShader: WebGLShader
  private fragmentShader: WebGLShader
  private program: WebGLProgram

  constructor() {
    this.gl = WebGL(1, 1, { alpha: false, depth: false, antialias: false })

    this.vertexShader = this.gl.createShader(this.gl.VERTEX_SHADER) as WebGLShader
    this.gl.shaderSource(this.vertexShader, `
    
    `)
    this.gl.compileShader(this.vertexShader)

    this.fragmentShader = this.gl.createShader(this.gl.FRAGMENT_SHADER) as WebGLShader
    this.gl.shaderSource(this.fragmentShader, `
    
    `)
    this.gl.compileShader(this.fragmentShader)

    this.program = this.gl.createProgram() as WebGLProgram
    this.gl.attachShader(this.program, this.vertexShader)
    this.gl.attachShader(this.program, this.fragmentShader)
    this.gl.linkProgram(this.program)
  }

  private standardGeometry: Float32Array = new Float32Array([
    -1.0, -1.0,
    1.0, -1.0,
    -1.0, 1.0,
    -1.0, 1.0,
    1.0, -1.0,
    1.0, 1.0
  ])

  private resizeGl(width: number, height: number) {
    const maxSize = this.gl.MAX_TEXTURE_SIZE

    if (width >= maxSize || height >= maxSize) {
      throw new Error('Maximum WebGL texture size exceeded')
    } else {
      this.gl.getExtension('STACKGL_resize_drawingbuffer').resize(Math.round(width), Math.round(height))
    }
  }

  private makeTexture(width: number, height: number, type: number, data: ArrayBufferView | null) {
    const gl = this.gl

    const texture = gl.createTexture()
    gl.bindTexture(gl.TEXTURE_2D, texture)

    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST)
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST)
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE)
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE)

    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, width, height, 0, gl.RGBA, type, data)

    gl.bindTexture(gl.TEXTURE_2D, null)

    return texture
  }

  private attatchFrameBuffer(texture: WebGLTexture) {
    const gl = this.gl

    const frameBuffer = gl.createFramebuffer()
    gl.bindFramebuffer(gl.FRAMEBUFFER, frameBuffer)
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, texture, 0)

    return frameBuffer
  }

  private compute(program: WebGLProgram) {
    // Create buffer containing vertex positions
    const vertexBuffer = this.gl.createBuffer() as WebGLBuffer
    this.gl.bindBuffer(this.gl.ARRAY_BUFFER, vertexBuffer)
    this.gl.bufferData(this.gl.ARRAY_BUFFER, this.standardGeometry, this.gl.STATIC_DRAW)

    // Load buffer into vertex buffer location
    const vertexBufferLocation = this.gl.getAttribLocation(program, 'position')
    this.gl.enableVertexAttribArray(vertexBufferLocation)
    this.gl.vertexAttribPointer(vertexBufferLocation, 2, this.gl.FLOAT, false, 0, 0)

    // Draw using program
    this.gl.useProgram(program)
    this.gl.drawArrays(this.gl.TRIANGLE_STRIP, 0, 4) // WHERE DOES THIS DRAW? & WHERE DO I NEED TO READ FROM?
  }

  private computeSimilarities(image: ImageData, template: ImageData) {
    // set dimensions
    const outputWidth = image.width-template.width+1
    const outputHeight = image.height-template.height+1
    const outputDepth = template.data.length/4
    this.resizeGl(outputWidth*outputDepth, outputHeight)

    // HOW DO I LOAD IN A TEXTURE?
    
    this.compute(this.program)
  }
}