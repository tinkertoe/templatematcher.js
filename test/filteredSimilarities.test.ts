import { TemplateMatcher } from '../src/index'
import { image, template } from './assets/sampleImageData'

test('filteredSimilarities', async () => {
  const tm = new TemplateMatcher()
  const result = await tm.filteredSimilarities(image, template, 51)
  expect(result).toMatchSnapshot()

  await tm.destroy()
})