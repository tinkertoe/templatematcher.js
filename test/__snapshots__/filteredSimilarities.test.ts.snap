// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filteredSimilarities 1`] = `
Array [
  Object {
    "allSimilarities": Float32Array [
      0,
      100,
      0,
      100,
      100,
      100,
    ],
    "averageSimilarity": 66.66667175292969,
    "location": Object {
      "x": 0,
      "y": 8,
    },
  },
  Object {
    "allSimilarities": Float32Array [
      100,
      100,
      100,
      100,
      100,
      100,
    ],
    "averageSimilarity": 100,
    "location": Object {
      "x": 1,
      "y": 9,
    },
  },
  Object {
    "allSimilarities": Float32Array [
      0,
      100,
      100,
      100,
      0,
      100,
    ],
    "averageSimilarity": 66.66667175292969,
    "location": Object {
      "x": 0,
      "y": 10,
    },
  },
]
`;
